core/trino-main/src/main/java/io/trino/sql/planner/EqualityInference.java

```diff
@@ -215,17 +214,34 @@ public EqualityPartition generateEqualitiesPartitionedBy(Set<Symbol> scope)
                         .forEach(scopeComplementEqualities::add);
             }
 
-            // Compile the scope straddling equality expressions
-            List<Expression> connectingExpressions = new ArrayList<>();
-            connectingExpressions.add(matchingCanonical);
-            connectingExpressions.add(complementCanonical);
-            connectingExpressions.addAll(scopeStraddlingExpressions);
-            connectingExpressions = connectingExpressions.stream()
-                    .filter(Objects::nonNull)
-                    .collect(Collectors.toList());
-            Expression connectingCanonical = getCanonical(connectingExpressions.stream());
+            // Compile single equality between matching and complement scope.
+            // Only consider expressions that don't have derived expression in other scope.
+            // Otherwise, redundant equality would be generated.
+            Optional<Expression> matchingConnecting = scopeExpressions.stream()
+                    .filter(expression -> SymbolsExtractor.extractAll(expression).isEmpty() || rewrite(expression, symbol -> !scope.contains(symbol), false) == null)
+                    .min(canonicalComparator);
+            Optional<Expression> complementConnecting = scopeComplementExpressions.stream()
+                    .filter(expression -> SymbolsExtractor.extractAll(expression).isEmpty() || rewrite(expression, scope::contains, false) == null)
+                    .min(canonicalComparator);
+            if (matchingConnecting.isPresent() && complementConnecting.isPresent() && !matchingConnecting.equals(complementConnecting)) {
+                scopeStraddlingEqualities.add(new ComparisonExpression(ComparisonExpression.Operator.EQUAL, matchingConnecting.get(), complementConnecting.get()));
+            }
+
+            // Compile the scope straddling equality expressions.
+            // scopeStraddlingExpressions couldn't be pushed to either side,
+            // therefore there needs to be an equality generated with
+            // one of the scopes (either matching or complement).
+            List<Expression> straddlingExpressions = new ArrayList<>();
+            if (matchingCanonical != null) {
+                straddlingExpressions.add(matchingCanonical);
+            }
+            else if (complementCanonical != null) {
+                straddlingExpressions.add(complementCanonical);
+            }
```

Comment:
What if both are non-null? What are the implications of adding one but not the other?

Reply:
So you have three groups of expressions:
```
scope
complement
straddling (these are guaranteed to have symbols from from scope and non-scope)
```
They are all equal.
Therefore you need to generate equalities
* within `scope`
* within `complement`
* within `straddling`
* between `scope<->complement`
AND
* between one of `(scope, complement) <-> straddling`

let's say:
```
scope = (l.x; l.y)
complement  = (r.x; r.y)
straddling = (l.i + r.j; l.m + r.n)
```
Then you generate:
```
for scope: l.x = l.y
for complement: r.x = r.y
for straddling: l.i + r.j = l.m + r.n
```
You also need to generate one canonical equality between scope and complement:
```
l.x = r.y
```
But you also need to connect `straddling` group with the rest somehow. The way you do it is by choosing canonical expression from EITHER `scope` or `complement` and generate equality with any straddling expression, e.g:
```
l.x = l.i + r.j 
```
This way yo generate all needed equalities, but not more.


If you were to choose canonical expression from both `scope` and `complement`, then you would generate redundant equality. Specifically this is important for constants, e.g:
```
equalitySet = [1, l.r, r.y, l.i + r.j]
scope = [1, l.r]
complement = [1, r.y]
straddling = [l.i + r.j]
```
generated equalities:
```
1 = l.r; 1 = r.y; 1 = l.i + r.j
```
If you were to choose canonical expression from both `scope` and `complement`, then it would generate equalities:
```
1 = l.r; 1 = r.y; 1 = 1; 1 = l.i + r.j
```
because `1` is canonical on both `scope` and `complement` side.